### ISSUE TITLE
Bug: Key Error when waiting for changeset/Max attempts exceeded

### ISSUE THREAD

**Issue by tirumerla** [aws/aws-sam-cli#4681]
### Description:

Hello Folks,

We are running into issues during "wait_for_changeset". Our changeset creation takes `~20 minutes` and we started seeing this issue from past 2 months mainly because we added more services as part of our nested stacks which increased the changeset creation time. Based on the previous bug reports here thought it was some syntactical issues in our templates but no issues were raised with `cfn lint` also changeset creation doesn't fail and will eventual gets to "CREATE_COMPLETE" so definitely not syntax issues in the templates.

This issue is mainly due to the [delay that was set](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/deploy/deployer.py#L311) to `5 seconds` and [assuming the changeset creation is quick](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/deploy/deployer.py#L310). With delay set to `5 seconds` and max attempts set to `120` if the changeset creation takes more than `> 10 mins` waiting on changeset will fail. 

I could be missing something/may be wrong here but that's what i'm thinking.

Prior to sam cli version `v1.72.0` we were seeing "Max attempts exceeded"

```
Waiting for changeset to be created..
Uploading to testing/5e2a0b45a2ccfed7bf77d88d96d85453.template  207618 / 207618  (100.00%)[4470] Failed to execute script '__main__' due to unhandled exception!

Traceback (most recent call last):
  File "samcli/lib/deploy/deployer.py", line 313, in wait_for_changeset
  File "botocore/waiter.py", line 54, in wait
  File "botocore/waiter.py", line 366, in wait
botocore.exceptions.WaiterError: Waiter ChangeSetCreateComplete failed: Max attempts exceeded

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "samcli/__main__.py", line 12, in <module>
  File "click/core.py", line 1130, in __call__
  File "click/core.py", line 1055, in main
  File "click/core.py", line 1657, in invoke
  File "click/core.py", line 1404, in invoke
  File "click/core.py", line 760, in invoke
  File "samcli/lib/cli_validation/image_repository_validation.py", line 92, in wrapped
  File "click/decorators.py", line 84, in new_func
  File "click/core.py", line 760, in invoke
  File "samcli/lib/telemetry/metric.py", line 175, in wrapped
  File "samcli/lib/telemetry/metric.py", line 150, in wrapped
  File "samcli/lib/utils/version_checker.py", line 41, in wrapped
  File "samcli/cli/main.py", line 92, in wrapper
  File "samcli/commands/_utils/cdk_support_decorators.py", line 38, in wrapped
  File "samcli/commands/deploy/command.py", line 221, in cli
  File "samcli/commands/deploy/command.py", line 359, in do_cli
  File "samcli/commands/deploy/deploy_context.py", line 180, in run
  File "samcli/commands/deploy/deploy_context.py", line 257, in deploy
  File "samcli/lib/deploy/deployer.py", line 528, in create_and_wait_for_changeset
  File "samcli/lib/deploy/deployer.py", line 318, in wait_for_changeset
KeyError: 'StatusReason'
error Command failed with exit code 1.

``` 

After upgrading to latest version `v1.72.0` we are now seeing "Key Error"

```
Error: 'StatusReason'
Traceback:
  File "click/core.py", line 1055, in main
  File "click/core.py", line 1657, in invoke
  File "click/core.py", line 1404, in invoke
  File "click/core.py", line 760, in invoke
  File "samcli/lib/cli_validation/image_repository_validation.py", line 92, in wrapped
  File "click/decorators.py", line 84, in new_func
  File "click/core.py", line 760, in invoke
  File "samcli/lib/telemetry/metric.py", line 183, in wrapped
  File "samcli/lib/telemetry/metric.py", line 150, in wrapped
  File "samcli/lib/utils/version_checker.py", line 41, in wrapped
  File "samcli/cli/main.py", line 92, in wrapper
  File "samcli/commands/_utils/cdk_support_decorators.py", line 38, in wrapped
  File "samcli/commands/deploy/command.py", line 221, in cli
  File "samcli/commands/deploy/command.py", line 359, in do_cli
  File "samcli/commands/deploy/deploy_context.py", line 180, in run
  File "samcli/commands/deploy/deploy_context.py", line 257, in deploy
  File "samcli/lib/deploy/deployer.py", line 528, in create_and_wait_for_changeset
  File "samcli/lib/deploy/deployer.py", line 318, in wait_for_changeset

An unexpected error was encountered while executing "sam deploy".
Search for an existing issue:
https://github.com/aws/aws-sam-cli/issues?q=is%3Aissue+is%3Aopen+Bug%3A%20sam%20deploy%20-%20KeyError
Or create a bug report:
https://github.com/aws/aws-sam-cli/issues/new?template=Bug_report.md&title=Bug%3A%20sam%20deploy%20-%20KeyError
error Command failed with exit code 1.
```

My assumption is both of the errors are related to same root cause that is there is not enough time for it to wait until changeset creation is complete.

### Workaround(s):

- Here's a snippet of a workaround we have been doing so far in github actions

```
     name: Botocore Exception Workaround
      if: steps.deploy-api.outcome == 'failure'
      shell: bash
      run: |
        STACKNAME='<stack_name>'
        STATUS=$(aws cloudformation list-change-sets  --stack-name $STACKNAME --query 'reverse(sort_by(Summaries,&CreationTime))[0].Status' --output text)
        CHANGESET_NAME=$(aws cloudformation list-change-sets  --stack-name $STACKNAME --query 'reverse(sort_by(Summaries,&CreationTime))[0].ChangeSetId' --output text)
        echo "Status: $STATUS, ChangeSet Name: $CHANGESET_NAME"
        if [[ "$STATUS" == "CREATE_IN_PROGRESS" ]]; then
          echo "ChangeSet is still in progress, waiting..."
          aws cloudformation wait change-set-create-complete --stack-name $STACKNAME --change-set-name $CHANGESET_NAME
          echo "list change set..."
          STATUS=$(aws cloudformation list-change-sets  --stack-name $STACKNAME --query 'reverse(sort_by(Summaries,&CreationTime))[0].Status' --output text)
        fi
        if [[ "$STATUS" == "CREATE_COMPLETE" ]]; then
          echo "Change set created successfully...."
          echo "Executing Changeset..."
          aws cloudformation execute-change-set --change-set-name $CHANGESET_NAME --stack-name $STACKNAME
          aws cloudformation wait stack-update-complete --stack-name $STACKNAME
          echo "Describing stack..."
          aws cloudformation describe-stacks --stack-name $STACKNAME --output table
          echo "Stack updated successfully...."
        else
          echo "Change set creation failed"
          exit 1
        fi
```

- I made a change to [Delay and MaxAttempts ](https://github.com/tirumerla/aws-sam-cli/blob/update-changeset-delay/samcli/lib/deploy/deployer.py#L311) and built the docker image locally to test it out and sure enough that works without any issues since we extended the wait period from `10mins` to `~1hr`.

### Steps to reproduce:

- Create a big enough sam template with multiple nested stacks that takes changeset creation `>10 mins` and `sam deploy`. You should see the "wait_for_changeset" to throw the same error.
- This is the command we are using 
   ```
    sam deploy \
  --stack-name "$STACKNAME" \
  --s3-bucket "$S3BUCKET" \
  --s3-prefix "$S3PREFIX" \
  --region "$AWS_REGION" \
  --capabilities CAPABILITY_IAM CAPABILITY_NAMED_IAM CAPABILITY_AUTO_EXPAND \
  --parameter-overrides "Stage=\"$STAGE\" SiteSubdomainName=\"$SUBDOMAIN\" SiteDomainName=\"$DOMAIN\" Elasticsearch=\"$ELASTICSEARCH\" $PARAMETER_OVERRIDES" \
  --no-fail-on-empty-changeset --debug;
   ```

### Observed result:

```
Waiting for changeset to be created..
2023-02-09 11:18:22,659 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2023-02-09 11:18:22,762 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2023-02-09 11:18:22,762 | Unable to find Click Context for getting session_id.
2023-02-09 11:18:22,763 | Sending Telemetry: {'metrics': [{'commandRun': {'requestId': '05d2bf62-7d5e-40ea-ac26-f11aa497c0f4', 'installationId': 'a77959f6-abd1-40f4-9d12-ebc8fab9773b', 'sessionId': 'b4e4f953-c300-4afe-a2f5-6ebe4100526b', 'executionEnvironment': 'CLI', 'ci': False, 'pyversion': '3.10.9', 'samcliVersion': '1.73.0', 'awsProfileProvided': False, 'debugFlagProvided': True, 'region': 'us-east-1', 'commandName': 'sam deploy', 'metricSpecificAttributes': {'projectType': 'CDK', 'gitOrigin': None, 'projectName': '7246d2b2b9ceb9cbbf5b0c7612f3258b97856f6bdb57a79d935308fcbf08270d', 'initialCommit': None}, 'duration': 761563, 'exitReason': 'KeyError', 'exitCode': 255}}]}
2023-02-09 11:18:22,763 | Sending Telemetry: {'metrics': [{'events': {'requestId': 'debe9c75-fd01-4168-96b6-91747b2a1c94', 'installationId': 'a77959f6-abd1-40f4-9d12-ebc8fab9773b', 'sessionId': 'b4e4f953-c300-4afe-a2f5-6ebe4100526b', 'executionEnvironment': 'CLI', 'ci': False, 'pyversion': '3.10.9', 'samcliVersion': '1.73.0', 'metricSpecificAttributes': {'events': [{'event_name': 'UsedFeature', 'event_value': 'CDK', 'thread_id': 8572134656, 'time_stamp': '2023-02-09 19:18:22.664'}]}}}]}
2023-02-09 11:18:22,992 | Telemetry response: 200
2023-02-09 11:18:22,997 | Telemetry response: 200

Error: 'StatusReason'
Traceback:
  File "/opt/homebrew/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/homebrew/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/homebrew/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/homebrew/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/lib/cli_validation/image_repository_validation.py", line 92, in wrapped
    return func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/click/decorators.py", line 84, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/lib/telemetry/metric.py", line 183, in wrapped
    raise exception  # pylint: disable=raising-bad-type
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/lib/telemetry/metric.py", line 150, in wrapped
    return_value = func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/lib/utils/version_checker.py", line 41, in wrapped
    actual_result = func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/cli/main.py", line 92, in wrapper
    return func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/commands/_utils/cdk_support_decorators.py", line 38, in wrapped
    return func(*args, **kwargs)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/commands/deploy/command.py", line 192, in cli
    do_cli(
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/commands/deploy/command.py", line 359, in do_cli
    deploy_context.run()
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/commands/deploy/deploy_context.py", line 166, in run
    return self.deploy(
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/commands/deploy/deploy_context.py", line 249, in deploy
    result, changeset_type = self.deployer.create_and_wait_for_changeset(
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/lib/deploy/deployer.py", line 528, in create_and_wait_for_changeset
    self.wait_for_changeset(result["Id"], stack_name)
  File "/opt/homebrew/lib/python3.10/site-packages/samcli/lib/deploy/deployer.py", line 318, in wait_for_changeset
    reason = resp["StatusReason"]

An unexpected error was encountered while executing "sam deploy".
Search for an existing issue:
https://github.com/aws/aws-sam-cli/issues?q=is%3Aissue+is%3Aopen+Bug%3A%20sam%20deploy%20-%20KeyError
Or create a bug report:
https://github.com/aws/aws-sam-cli/issues/new?template=Bug_report.md&title=Bug%3A%20sam%20deploy%20-%20KeyError
```

### Expected result:

- wait_for_changeset to wait without failing until "CREATE_COMPLETE" status.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: Tested on `Mac` ( `OS X ventura 13.0.1` ) & also on linux ( `ubuntu 22.04` )
2. `sam --version`: `v1.73.0`
3. AWS region: `us-east-1`

```
# Paste the output of `sam --info` here
```
```
{
  "version": "1.73.0",
  "system": {
    "python": "3.10.9",
    "os": "macOS-13.0.1-arm64-arm-64bit"
  },
  "additional_dependencies": {
    "docker_engine": "20.10.11",
    "aws_cdk": "Not available",
    "terraform": "1.3.4"
  }
}
```

**Comment by jfuss** [comment]
@tirumerla Thanks for the report. I didn't know changesets could take so long :) 

We added `SAM_CLI_POLL_DELAY` to `sam deploy` which did something similar but targeted the part of deploy that fetches updates. https://github.com/aws/aws-sam-cli/blob/6dc1a5666b8662734cb2aca8581ed11444d724cf/samcli/commands/deploy/command.py#L325 and https://github.com/aws/aws-sam-cli/blob/6dc1a5666b8662734cb2aca8581ed11444d724cf/samcli/lib/deploy/deployer.py#L383 are some relevant parts.

I think what we should do here is either re-use that Env Var or create one targeted for the changeset call. Reusing is probably reasonable here but not sure if it would be important for SAM CLI to control those independently.

Do you think that would give you the ability to solve the issue you are seeing? Want to make sure it would before we go further.

**Comment by tirumerla** [comment]
> We added `SAM_CLI_POLL_DELAY` to `sam deploy` which did something similar but targeted the part of deploy that fetches updates.

Thanks @jfuss for the info. This is exactly what i have tried to get around this issue initally by adding these env variables....after looking at the code i realized it was only for describe stacks API calls :) 

> I think what we should do here is either re-use that Env Var or create one targeted for the changeset call. Reusing is probably reasonable here but not sure if it would be important for SAM CLI to control those independently.

> Do you think that would give you the ability to solve the issue you are seeing? Want to make sure it would before we go further.

Sorry yeah I forgot to mention we could definitely extend/re-use these env. variables to wait for changeset call which should help us with this issue!

**Comment by jfuss** [comment]
@tirumerla I chatted with the team yesterday. I think re-using the same logic for the describe-stack call with `SAM_CLI_POLL_DELAY`.

Would you be interested in contributing it? The main thing would port the same logic to the wrap the create-change-set call and then some tests for the change.

**Comment by tirumerla** [comment]
> @tirumerla I chatted with the team yesterday. I think re-using the same logic for the describe-stack call with `SAM_CLI_POLL_DELAY`.
> 
> Would you be interested in contributing it? The main thing would port the same logic to the wrap the create-change-set call and then some tests for the change.

Sure i can take a look. Thanks for bringing this up with the team. 

Do you think [this ](https://github.com/aws/aws-sam-cli/compare/develop...tirumerla:aws-sam-cli:poll-delay-changesets?expand=1)should be enough?

**Comment by jfuss** [comment]
@tirumerla That looks reasonable to me. I assume you have tried it locally and resolves your issue as well? 

I think the only thing we will want is some unit testing that can validate we are setting the right value of `client_sleep` within the boto3 client. Something within this test class seems reasonable to me: https://github.com/aws/aws-sam-cli/blob/develop/tests/unit/lib/deploy/test_deployer.py#L62

**Comment by tirumerla** [comment]
> I assume you have tried it locally and resolves your issue as well?]

@jfuss that is correct.

Raised the PR, let me know what you think.

**Comment by hawflau** [comment]
Closing as the fix has been released in AWS SAM CLI v1.77.0

**Comment by rmaafs** [comment]
> @tirumerla Thanks for the report. I didn't know changesets could take so long :)
> 
> We added `SAM_CLI_POLL_DELAY` to `sam deploy` which did something similar but targeted the part of deploy that fetches updates.
> 
> https://github.com/aws/aws-sam-cli/blob/6dc1a5666b8662734cb2aca8581ed11444d724cf/samcli/commands/deploy/command.py#L325
> 
> and
> https://github.com/aws/aws-sam-cli/blob/6dc1a5666b8662734cb2aca8581ed11444d724cf/samcli/lib/deploy/deployer.py#L383
> 
> are some relevant parts.
> I think what we should do here is either re-use that Env Var or create one targeted for the changeset call. Reusing is probably reasonable here but not sure if it would be important for SAM CLI to control those independently.
> 
> Do you think that would give you the ability to solve the issue you are seeing? Want to make sure it would before we go further.

From my side, I just created this environment variable on my virtual machine and works! Now my deployment is successfully.
Commands:
```
echo "SAM_CLI_POLL_DELAY=300" >> $GITHUB_ENV
export SAM_CLI_POLL_DELAY=300
```
